Backend readme improvements (#110)
teamhanko/hanko: backend/README.md

    origin: <CHANGE-ME>
```

> **Note** You need to change the smtp config to start the service. You can enter any host, user and password,
> they will not be checked for correctness at startup. But be aware that no emails will be sent
> and your users might not be able to login if no valid smtp server is set up.

> **Note** `secrets.keys` must be a random generated string at least 16 characters long.

### Docker

#### Database migrations

Before you can start and use the service you need to run the database migrations:

```shell
docker run --mount type=bind,source=<PATH-TO-CONFIG-FILE>,target=/config/config.yaml -p 8000:8000 -it ghcr.io/teamhanko/hanko:main migrate up
```

> **Note** The `<PATH-TO-CONFIG-FILE>` must be an absolute path to your config file created above.

#### Start the service

To start the service just run:

```shell
docker run --mount type=bind,source=<PATH-TO-CONFIG-FILE>,target=/config/config.yaml -p 8000:8000 -it ghcr.io/teamhanko/hanko:main serve public
```

> **Note** The `<PATH-TO-CONFIG-FILE>` must be an absolute path to your config file created above.

The service is now available at `localhost:8000`.

### From source

#### Building

> To build the Hanko backend you only need to have [go installed](https://go.dev/doc/install) installed on your computer.

Reviewer: ```suggestion
To build the Hanko backend you only need to have [go installed](https://go.dev/doc/install) on your computer.
```

# Hanko backend

Hanko backend provides an HTTP API to build a modern login and registration experience for your users. Its core features are an API for passkeys (WebAuthn), passwords, and passcodes, as well as JWT management.

Hanko backend can be used on its own or in combination with [hanko-js](../hanko-js), a powerful frontend library that contains polished and customizable UI flows for password-based and passwordless user authentication that can be easily integrated into any web app with as little as two lines of code.

## API features

- Passkeys (WebAuthn)
- Passcodes
- Passwords
- Email verification
- JWT management

## Upcoming features

- Exponential backoff for password attempts and passcode email sending
- More tests and code documentation

## Basic usage

The easiest way to start Hanko backend service is through docker. But before we can do that, we need to create a config file.

> > **Note** If you just want to jump right into the experience of passkeys and passcodes, head over to the [quickstart guide](../README.md#Quickstart).

Reviewer: > [...] the [quickstart guide](../README.md#Quickstart).

There is no heading (and section/guide) with that ID it just takes me to the top of the readme.

Reviewer 2: Any idea how this can be fixed without using the absolute url?

Reviewer: Wel, you could try using the "correct" ID for a start, so maybe:

```suggestion
> **Note** If you just want to jump right into the experience of passkeys and passcodes, head over to the [quickstart guide](../README.md#quickstart).
```

Reviewer: Nope, does not work. I'll just try something else real quick.

# Hanko backend

> The Hanko backend provides public and private APIs that offer a set of functions to create a modern login and registration experience for your users. Hanko backend works well with [hanko-js](../hanko-js), a web component that can easily be integrated in any web app and provides ready-to-use login and registration UI flows.

Reviewer 3: hanko-js is not a web component, but among other things you can use hanko-js to register a web component.

Reviewer 3: A small suggestion here would be not to put the web component in the center, but to say that the backend consists of an API for passkeys/webauthn, passwords and passcodes, and manages JWT's.

Reviewer 3: Hmm, I read through it again. About the upcoming features we say that backoff mechanisms will come, but at the bottom of the text it says:

> The Hanko service does not implement rate limiting in any way. In production systems, you want to hide the Hanko service behind a proxy or gateway (e.g. kong, traeffik) that implements rate limiting.
 
And exponential backoff is is a rate limiting algorithm.

Reviewer 2: Thanks for your feedback. How about that:

```suggestion
Hanko backend provides public and private APIs that offer a set of functions to create a modern login and registration experience for your users. Its core features are an API for passkeys (WebAuthn), passwords, and passcodes, as well as JWT management.

Hanko backend can be used on its own or in combination with [hanko-js](../hanko-js), a powerful frontend library that contains polished and customizable UI flows for password-based and passwordless user authentication that can be easily integrated into any web app with as little as two lines of code.
```

Reviewer 2: > Hmm, I read through it again. About the upcoming features we say that backoff mechanisms will come, but at the bottom of the text it says:
> 
> > The Hanko service does not implement rate limiting in any way. In production systems, you want to hide the Hanko service behind a proxy or gateway (e.g. kong, traeffik) that implements rate limiting.
> 
> And exponential backoff is is a rate limiting algorithm. I would suggest either taking it out of the list of upcoming features or being more precise in the text, or the other way around, depending on how we want it to work later.

Addressed this here https://github.com/teamhanko/hanko/pull/110/commits/91fa76d4805ee1670b8ef6f7583faf7975ec5f8b
